client-sdk/ts-web/rt/src/callformat.ts

```diff
@@ -1,4 +1,146 @@
+import * as randomBytes from 'randombytes';
+
+import * as mraeDeoxysii from './mrae/deoxysii';
+import * as transaction from './transaction';
+import * as types from './types';
+
 /**
  * Call data key pair ID domain separation context base.
  */
 export const CALL_DATA_KEY_PAIR_ID_CONTEXT_BASE = 'oasis-runtime-sdk/private: tx';
+
+/**
+ *  EncodeConfig is call encoding configuration.
+ *  golang: oasis-sdk/client-sdk/go/callformat/callformat.go
+ *  rust:
+ */
+export interface EncodeConfig {
+    /**
+     * publicKey is an optional runtime's call data public key to use for encrypted call formats.
+     */
+    publicKey?: types.KeyManagerSignedPublicKey;
+}
+
+export interface MetaEncryptedX25519DeoxysII {
+    sk: Uint8Array;
+    pk: Uint8Array;
+}
+
+/**
+ * encodeCallWithNonceAndKeys encodes a call based on its configured call format.
+ * It returns the encoded call and any metadata needed to successfully decode the result.
+ */
+export async function encodeCallWithNonceAndKeys(
+    nonce: Uint8Array,
+    sk: Uint8Array,
+    pk: Uint8Array,
+    call: types.Call,
+    format: types.CallFormat,
+    config?: EncodeConfig,
+): Promise<[types.Call, unknown]> {
+    switch (format) {
+        case transaction.CALLFORMAT_PLAIN:
+            return [call, undefined];
+        case transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII:
+            if (config?.publicKey === undefined) {
+                throw new Error('callformat: runtime call data public key not set');
+            }
+            const rawCall = oasis.misc.toCBOR(call);
+            const zeroBuffer = new Uint8Array(0);
+            const sealedCall = mraeDeoxysii.boxSeal(
+                nonce,
+                rawCall,
+                zeroBuffer,
+                config.publicKey.key,
+                sk,
+            );
+            const envelope: types.CallEnvelopeX25519DeoxysII = {
+                pk: pk,
+                nonce: nonce,
+                data: sealedCall,
+            };
+            const encoded: types.Call = {
+                format: transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII,
+                method: '',
+                body: oasis.misc.toCBOR(envelope),
+            };
+            const meta: MetaEncryptedX25519DeoxysII = {
+                sk: sk,
+                pk: config.publicKey.key,
+            };
+            return [encoded, meta];
+        default:
+            throw new Error(`callformat: unsupported call format: ${format}`);
+    }
+}
+
+/**
+ * encodeCall randomly generates nonce and keyPair and then call encodeCallWithNonceAndKeys
+ * It returns the encoded call and any metadata needed to successfully decode the result.
+ */
+export async function encodeCall(
+    call: types.Call,
+    format: types.CallFormat,
+    config?: EncodeConfig,
+): Promise<[types.Call, unknown]> {
+    const nonce = randomBytes(deoxysii.NonceSize);
+    const keyPair = nacl.box.keyPair();
+    return await encodeCallWithNonceAndKeys(
```

Comment:
do I need to use `await` here? I think I can directly return the Promise, right?